src/protagonist/API/Features/OriginStrategies/Requests/UpdateCustomerOriginStrategy.cs

```diff
@@ -0,0 +1,127 @@
+    {
+        this.dbContext = dbContext;
+        this.credentialsExporter = credentialsExporter;
+    }
+
+    public async Task<ModifyEntityResult<CustomerOriginStrategy>> Handle(
+        UpdateCustomerOriginStrategy request,
+        CancellationToken cancellationToken)
+    {
+        var wipeCredentials = false;
+        var existingStrategy = await dbContext.CustomerOriginStrategies.SingleOrDefaultAsync(
+            s => s.Id == request.StrategyId && s.Customer == request.CustomerId,
+            cancellationToken);
+        
+        if (existingStrategy == null)
+            return ModifyEntityResult<CustomerOriginStrategy>
+                .Failure($"Couldn't find an origin strategy with the id {request.StrategyId}", WriteResult.NotFound);
+        
+        if (!string.IsNullOrWhiteSpace(request.Regex))
+        {
+            var regexUsed = await dbContext.CustomerOriginStrategies.AnyAsync(
+                s => s.Customer == request.CustomerId && s.Regex == request.Regex && s.Id != existingStrategy.Id,
+                cancellationToken);
+
+            if (regexUsed)
+                return ModifyEntityResult<CustomerOriginStrategy>.Failure(
+                    "An origin strategy using the same regex already exists",
+                    WriteResult.Conflict);
+
+            existingStrategy.Regex = request.Regex;
+        }
+
+        if (request.Strategy.HasValue)
+        {
+            if(existingStrategy.Strategy == OriginStrategyType.BasicHttp && request.Strategy != OriginStrategyType.BasicHttp
+                && !string.IsNullOrWhiteSpace(existingStrategy.Credentials))
+                wipeCredentials = true;
+            existingStrategy.Strategy = request.Strategy.Value;
+        }
+        
+        if (request.Optimised.HasValue)
+            existingStrategy.Optimised = request.Optimised.Value;
+        
+        if (request.Order.HasValue)
+            existingStrategy.Order = request.Order.Value;
+        
+        if(existingStrategy.Strategy == OriginStrategyType.BasicHttp && string.IsNullOrWhiteSpace(request.Credentials)
+           && (request.Strategy.HasValue || request.Credentials.HasText()))
+            return ModifyEntityResult<CustomerOriginStrategy>
+                .Failure("Credentials must be specified when using basic-http-authentication as an origin strategy", WriteResult.FailedValidation);
+        
+        if (existingStrategy.Strategy != OriginStrategyType.S3Ambient && existingStrategy.Optimised
+            && (request.Strategy.HasValue || request.Optimised.HasValue))
+            return ModifyEntityResult<CustomerOriginStrategy>
+                .Failure("'Optimised' is only applicable when using s3-ambient as an origin strategy", WriteResult.FailedValidation);
+        
+        if (!string.IsNullOrWhiteSpace(request.Credentials))
+        {
+            if (!(!string.IsNullOrEmpty(request.Regex) && request.Strategy.HasValue && request.Optimised.HasValue && request.Order.HasValue))
```

Comment:
The double negative at the start here is difficult to read, is this a case for `request.Regex.HasText()`? Also might read easier if it was a named method, `if (!isFullOriginStrategy(request))`